File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()

        n = len(nums)
        answer = []
        for i, num in enumerate(nums):
            if i > 0 and nums[i - 1] == num:
                continue
            l = i + 1
            r = n - 1
            while l < r:
                sum = num + nums[l] + nums[r]
                if sum == 0:
                    answer.append([num, nums[l], nums[r]])
                    while l < n - 1 and nums[l] == nums[l + 1]:
                        l += 1
                    while r > 0 and nums[r - 1] == nums[r]:
                        r -= 1
                    l += 1
                    r -= 1
                elif sum > 0:
                    r -= 1
                else:
                    l += 1

        return answer

File: 15-3sum/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        int n = nums.length;
        Arrays.sort(nums);
        List<List<Integer>> answer = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            if (i > 0 && nums[i-1] == nums[i])
                continue;
            int l = i + 1;
            int r = n - 1;
            while (l < r) {
                int target = nums[l] + nums[i] + nums[r];
                if (target == 0) {
                    answer.add(new ArrayList<>(Arrays.asList(nums[l++], nums[i], nums[r--])));
                    while (l < n - 1 && nums[l-1] == nums[l]) {
                        l++;
                    }
                    while (l < r && nums[r + 1] == nums[r]) {
                        r -= 1;
                    }
                } else if (target > 0) {
                    r -= 1;
                } else {
                    l += 1;
                }
            }
        }
        return answer;
    }
}